.circleci/patch_log_device.rb

```diff
@@ -0,0 +1,16 @@
+if RUBY_ENGINE == 'jruby'
+  require 'logger'
+  ::Logger::LogDevice.prepend(
+    Module.new do
+      def write(message)
+        $stdout.puts("[Writing] ProcessID: #{Process.pid}, ThreadID: #{Thread.current.object_id}, Message: #{message}")
+        super(message)
+      end
+
+      def close
+        $stdout.puts("[Closing] ProcessID: #{Process.pid}, ThreadID: #{Thread.current.object_id}")
+        super
+      end
+    end
```

Comment:
I would suggest using the same `warn(...)` mechanism that the logger does, because we're not actually sure that `$stdout` is open at this time -- it may have been closed.

Also, I'd suggest printing every thread (`Thread.list`) and their backtrace, so you'll know what may be staying over and perhaps making the app not quit.

Also, instrumenting every `write` and `close` may get a bit slow/chatty? I'd suggest instrumenting only when there's an exception, e.g. when we know the issue has been triggered.

Reply:
`warn` should also output to standard error rather than standard output which I think would be more correct?

Reply:
- `$stdout` or `STDOUT` may also be taken over (say by some mock or stub). `File.open(/dev/stdout)`? (or err)?
- either or both of `$stdout`/`err` may or may not be closed. Might be worth querying and logging their state via `$stdout.closed?` as well?
- what about logging to a _file_?

> Also, instrumenting every write and close may get a bit slow/chatty?

Let's get some (any) output first and see if we're having some patterns emerge.

> instrumenting only when there's an exception, e.g. when we know the issue has been triggered.

We dont' know what's causing it, or do we? Seems speculative to tie log output to a hypothetical.